src/protocol/oprf/mod.rs

```diff
@@ -0,0 +1,330 @@
+    Ok(res)
+}
+
+// --------------------------------------------------------------------------- //
+
+fn split_shares(
+    input_rows: &[OPRFInputRow],
+    direction: Direction,
+) -> impl Iterator<Item = OPRFShare> + '_ {
+    let f = move |input_row| OPRFShare::from_input_row(input_row, direction);
+    input_rows.iter().map(f)
+}
+
+fn combine_shares<L, R>(l: L, r: R) -> Vec<OPRFInputRow>
+where
+    L: IntoIterator<Item = OPRFShare>,
+    R: IntoIterator<Item = OPRFShare>,
+{
+    l.into_iter()
+        .zip(r)
+        .map(|(l, r)| l.to_input_row(r))
+        .collect::<Vec<_>>()
+}
+
+fn add_single_shares<'i, T, L, R>(l: L, r: R) -> impl Iterator<Item = T::Output> + 'i
+where
+    T: Add,
+    L: IntoIterator<Item = T> + 'i,
+    R: IntoIterator<Item = T> + 'i,
+{
+    l.into_iter().zip(r).map(|(a, b)| a + b)
+}
+
+// --------------------------------------------------------------------------- //
+
+fn generate_random_tables_with_peers<C: Context>(
+    batch_size: u32,
+    narrow_ctx: &C,
+) -> (
+    impl Iterator<Item = OPRFShare> + '_,
+    impl Iterator<Item = OPRFShare> + '_,
+) {
+    let (rng_l, rng_r) = narrow_ctx.prss_rng();
+    let with_left = sample_iter(rng_l).take(batch_size as usize);
+    let with_right = sample_iter(rng_r).take(batch_size as usize);
+    (with_left, with_right)
+}
+
+fn generate_random_table_solo<C>(
+    batch_size: u32,
+    narrow_ctx: &C,
+    peer: Direction,
+) -> impl Iterator<Item = OPRFShare> + '_
+where
+    C: Context,
+{
+    let rngs = narrow_ctx.prss_rng();
+    let rng = match peer {
+        Direction::Left => rngs.0,
+        Direction::Right => rngs.1,
+    };
+
+    sample_iter(rng).take(batch_size as usize)
+}
+
+fn sample_iter<R: Rng>(rng: R) -> impl Iterator<Item = OPRFShare> {
+    rng.sample_iter(Standard)
+}
+
+// ---------------------------- helper communication ------------------------------------ //
+
+async fn send_to_peer<C: Context, I: IntoIterator<Item = OPRFShare>>(
+    ctx: &C,
+    step: &OPRFShuffleStep,
+    direction: Direction,
+    items: I,
+) -> Result<(), Error> {
+    let role = ctx.role().peer(direction);
+    let send_channel = ctx.narrow(step).send_channel(role);
+    for (record_id, row) in items.into_iter().enumerate() {
+        send_channel.send(RecordId::from(record_id), row).await?;
+    }
+    Ok(())
+}
+
+async fn receive_from_peer<C: Context>(
+    ctx: &C,
+    step: &OPRFShuffleStep,
+    direction: Direction,
+    batch_size: u32,
+) -> Result<Vec<OPRFShare>, Error> {
+    let role = ctx.role().peer(direction);
+    let receive_channel: ReceivingEnd<OPRFShare> = ctx.narrow(step).recv_channel(role);
+
+    let mut output: Vec<OPRFShare> = Vec::with_capacity(batch_size as usize);
+    for record_id in 0..batch_size {
+        let msg = receive_channel.receive(RecordId::from(record_id)).await?;
+        output.push(msg);
+    }
+
+    Ok(output)
+}
+
+// ------------------ Pseudorandom permutations functions -------------------- //
+
```

Comment:
similar as above, do we need separate functions here? I think it makes the code less clean.